File: frontend/src/dotGenerator.js
import { DOT_SHAPES } from './constants.js'

function qid(s) {
  return /^[a-zA-Z_]\w*$/.test(s) ? s : `"${s.replace(/\\/g, '\\\\').replace(/"/g, '\\"')}"`
}

function fmtAttrs(obj) {
  const pairs = Object.entries(obj)
    .map(([k, v]) => `${k}="${String(v).replace(/\\/g, '\\\\').replace(/"/g, '\\"')}"`)
    .join(', ')
  return `[${pairs}]`
}

export function generateDot({ name, goal, stylesheet, nodes, edges }) {
  const graphName = name || 'Pipeline'
  const lines = [`digraph ${qid(graphName)} {`]

  const ga = {}
  if (goal) ga.goal = goal
  if (stylesheet) ga.model_stylesheet = stylesheet
  if (Object.keys(ga).length) lines.push(`    graph ${fmtAttrs(ga)}`)
  lines.push('')

  nodes.forEach((d) => {
    const a = { shape: DOT_SHAPES[d.nodeType] || d.nodeType }
    if (d.label && d.label !== d.id) a.label = d.label
    if (d.prompt) a.prompt = d.prompt
    if (d.nodeType === 'manager') a.max_cycles = String(d.max_cycles != null ? d.max_cycles : 3)
    if (d.mcp_servers) a.mcp_servers = d.mcp_servers
    if (d.command) a.command = d.command
    if (d.http_url) a.url = d.http_url
    if (d.http_method && d.http_method !== 'GET') a.method = d.http_method
    if (d.http_body) a.body = d.http_body
    if (d.http_headers) a.headers = d.http_headers
    if (d.goal_gate) a.goal_gate = 'true'
    if (d.max_retries > 0) a.max_retries = String(d.max_retries)
    if (d.retry_target) a.retry_target = d.retry_target
    if (d.timeout) a.timeout = d.timeout
    if (d.llm_model) a.llm_model = d.llm_model
    lines.push(`    ${qid(d.id)} ${fmtAttrs(a)}`)
  })

  lines.push('')

  edges.forEach((d) => {
    const a = {}
    if (d.label) a.label = d.label
    if (d.condition) a.condition = d.condition
    if (d.weight > 0) a.weight = String(d.weight)
    if (d.fidelity) a.fidelity = d.fidelity
    if (d.loop_restart) a.loop_restart = 'true'
    const suffix = Object.keys(a).length ? ' ' + fmtAttrs(a) : ''
    lines.push(`    ${qid(d.source)} -> ${qid(d.target)}${suffix}`)
  })

  lines.push('}')
  return lines.join('\n')
}

File: attractor/server/dot_generator.py
"""Generate DOT source from a visual graph definition."""

from __future__ import annotations

from typing import Any

SHAPE_MAP = {
    "start": "Mdiamond",
    "exit": "Msquare",
    "codergen": "box",
    "conditional": "diamond",
    "human": "house",
    "tool": "parallelogram",
    "parallel": "component",
    "fan_in": "tripleoctagon",
    "manager": "hexagon",
}


def generate_dot(
    nodes: list[dict[str, Any]],
    edges: list[dict[str, Any]],
    graph_attrs: dict[str, Any] | None = None,
) -> str:
    lines: list[str] = []
    name = (graph_attrs or {}).get("name", "Pipeline")
    lines.append(f"digraph {_quote_id(name)} {{")

    g_attrs = dict(graph_attrs or {})
    g_attrs.pop("name", None)
    if g_attrs:
        attr_str = _format_attrs(g_attrs)
        lines.append(f"    graph {attr_str}")

    lines.append("")

    for node in nodes:
        node_id = node.get("id", "")
        if not node_id:
            continue

        attrs: dict[str, str] = {}
        node_type = node.get("type", "codergen")

        shape = SHAPE_MAP.get(node_type, node_type)
        attrs["shape"] = shape

        if node.get("label") and node["label"] != node_id:
            attrs["label"] = node["label"]
        if node.get("prompt"):
            attrs["prompt"] = node["prompt"]
        if node.get("goal_gate"):
            attrs["goal_gate"] = "true"
        if node.get("max_retries"):
            attrs["max_retries"] = str(node["max_retries"])
        if node.get("retry_target"):
            attrs["retry_target"] = node["retry_target"]
        if node.get("timeout"):
            attrs["timeout"] = node["timeout"]
        if node.get("command"):
            attrs["command"] = node["command"]
        if node.get("llm_model"):
            attrs["llm_model"] = node["llm_model"]
        if node.get("css_class"):
            attrs["class"] = node["css_class"]

        for k, v in node.get("extra_attrs", {}).items():
            attrs[k] = str(v)

        attr_str = _format_attrs(attrs) if attrs else ""
        lines.append(f"    {_quote_id(node_id)} {attr_str}")

    lines.append("")

    for edge in edges:
        source = edge.get("source", "")
        target = edge.get("target", "")
        if not source or not target:
            continue

        attrs: dict[str, str] = {}
        if edge.get("label"):
            attrs["label"] = edge["label"]
        if edge.get("condition"):
            attrs["condition"] = edge["condition"]
        if edge.get("weight"):
            attrs["weight"] = str(edge["weight"])
        if edge.get("fidelity"):
            attrs["fidelity"] = edge["fidelity"]
        if edge.get("loop_restart"):
            attrs["loop_restart"] = "true"

        for k, v in edge.get("extra_attrs", {}).items():
            attrs[k] = str(v)

        attr_str = f" {_format_attrs(attrs)}" if attrs else ""
        lines.append(
            f"    {_quote_id(source)} -> {_quote_id(target)}{attr_str}"
        )

    lines.append("}")
    return "\n".join(lines)


def _quote_id(name: str) -> str:
    if name.isidentifier() and name.isascii():
        return name
    return f'"{name}"'


def _format_attrs(attrs: dict[str, str]) -> str:
    parts: list[str] = []
    for k, v in attrs.items():
        parts.append(f'{k}="{_escape(v)}"')
    return "[" + ", ".join(parts) + "]"


def _escape(s: str) -> str:
    return s.replace("\\", "\\\\").replace('"', '\\"')
